### PR TITLE
sqlproxyccl: add a healthcheck and metrics

### PR DESCRIPTION
**Pull request by chrisseto**
Previous the SQL Proxy exposed no metrics nor any healthcheck.

For the SQL Proxy to function optimally when deployed in a Kubernetes
cluster, a liveness probe needs to be available to automatically detect
any issues.
For SREs to manage the SQL Proxy in a production environment, it must
provide some amount of visibility into its operations, namely health
checks and metrics.
This commits adds a rudimentary heathcheck endpoint at /_status/healthz/
and minimal metrics at /_status/vars/. More importantly, it adds the
plumbing required to easily add more metrics and more intelligent
health checks in the future.

Release note: None

As a side note, I'm not entirely certain of the vision for the proxy. It may make sense to instead build this functionality else where and just provide hooks in the proxy itself?
I _think_ it makes sense for metrics to live in the proxy itself as most systems that deploy it will want to monitor its health in some fashion.

### CODE REVIEW

**Comment by tbg** [comment]
I haven't scrutinized the code changes too closely, but this all looks good. I'm +1 on adding this stuff in sqlproxyccl (rather than on the CC side) in general. There are also plans on the SQL side to build upon the proxy code, see internal doc [here](https://docs.google.com/document/d/1abtfl-RpAeZCeHOtPC_NpeNc0yy1ggxpwDgm61R8Pt0/edit?ts=5f6e2cf8). I'll defer to them for a real review. cc @knz

File: pkg/ccl/sqlproxyccl/server.go
Comment (knz) [inline pkg/ccl/sqlproxyccl/server.go:58]: Can you explain why it's called "healthz" and not "health"

(also why `/_status` - is that standard?)

**Review by knz — APPROVED** [review]
LGTM

> @tbg There are also plans on the SQL side to build upon the proxy code, see internal doc here. 

As far as that doc was intended, this work aims to reuse the mtproxy code. So Chris' change here is congruent.

**Comment by tbg** [comment]
https://stackoverflow.com/questions/43380939/where-does-the-convention-of-using-healthz-for-application-health-checks-come-f

File: pkg/ccl/sqlproxyccl/server.go
Comment (chrisseto) [reply pkg/ccl/sqlproxyccl/server.go:61]: I was actually just copying the endpoints from CRDB 😅 

Most of cloud's auxiliary services expose metrics at the same endpoint as well.

**Comment by chrisseto** [comment]
TFTR! I've updated this PR with some additional metrics inside of the `Proxy` function itself and (hopefully) fixed CI failures.

**Review by imjching — COMMENTED** [review]
> Reviewed 6 of 6 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/55430)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andy-kimball, @chrisseto, and @knz)

___
*[pkg/ccl/sqlproxyccl/metrics.go, line 29 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/55430#-MJZaTT-9leGh1qmdqe-:-MJZaTT-9leGh1qmdqe0:b-ykd052) ([raw file](https://github.com/cockroachdb/cockroach/blob/f1619dd4f27e6dd14d7e2903a1ffd866c238e86d/pkg/ccl/sqlproxyccl/metrics.go#L29)):*
> ```
> metaCurConnCount
> ```

nit: How about calling this `metaActiveConnCount`?
___
*[pkg/ccl/sqlproxyccl/metrics.go, line 35 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/55430#-MJZaF_BEATnYYKV27Sa:-MJZaF_BEATnYYKV27Sb:b-a27726) ([raw file](https://github.com/cockroachdb/cockroach/blob/f1619dd4f27e6dd14d7e2903a1ffd866c238e86d/pkg/ccl/sqlproxyccl/metrics.go#L35)):*
> ```
> metaRoutingErrCount
> ```

nit: `s/metaRoutingErrCount/metaRoutingFailedCount`
___
*[pkg/ccl/sqlproxyccl/metrics.go, line 54 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/55430#-MJZa-qA400eDgRfM_6b:-MJZa-qB6jqqWoNTD0l1:b-vc0tfx) ([raw file](https://github.com/cockroachdb/cockroach/blob/f1619dd4f27e6dd14d7e2903a1ffd866c238e86d/pkg/ccl/sqlproxyccl/metrics.go#L54)):*
> ```
> Name:        "proxy.sql.backenddisconnect",
> ```
Why is this metric name the same as `metaBackendDisconnectCount`?

nit: could we also use underscores to split the words, e.g. routing_failed, backend_down, etc?
___
*[pkg/ccl/sqlproxyccl/server.go, line 48 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/55430#-MJZb69vCHIDAo5q7jGa:-MJZb69vCHIDAo5q7jGb:bkpkg2m) ([raw file](https://github.com/cockroachdb/cockroach/blob/f1619dd4f27e6dd14d7e2903a1ffd866c238e86d/pkg/ccl/sqlproxyccl/server.go#L48)):*
> ```
> registry.AddMetricStruct(proxyMetrics)
> ```

Not needed in this PR, but I'd imagine we want to expose `registry` one day, or at least some form of `AddMetricStruct`. This allows custom implementations, e.g. mtproxy in the context of cockroachdb/cockroach, and sqlproxy in the context of CC to specify its own metrics assuming it does more stuff other than just a regular proxying. For example, sqlproxy could talk to Kubernetes to grab the right routing rule, and maybe that would be a good place to collect metrics as well.
___
*[pkg/ccl/sqlproxyccl/server.go, line 91 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/55430#-MJZcnnK8JI327qb_Eg1:-MJZcnnK8JI327qb_Eg2:bcr8tx0) ([raw file](https://github.com/cockroachdb/cockroach/blob/f1619dd4f27e6dd14d7e2903a1ffd866c238e86d/pkg/ccl/sqlproxyccl/server.go#L91)):*
> ```Go
> 
> 	w.Header().Set(httputil.ContentTypeHeader, httputil.PlaintextContentType)
> 	s.prometheusExporter.ScrapeRegistry(s.metricsRegistry, true)
> ```

nit: `s.prometheusExporter.ScrapeRegistry(s.metricsRegistry, true /* includeChildMetrics*/)`

**Review by knz — COMMENTED** [review]
No more comments from me beyond what Jay already wrote. LGTM

> Reviewed 6 of 6 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/55430#-:-MJbFRhs9vA-KRKp_CpD:b-p3zcur)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andy-kimball, @chrisseto, and @knz)

**Review by chrisseto — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/55430)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andy-kimball, @imjching, and @knz)

___
*[pkg/ccl/sqlproxyccl/metrics.go, line 29 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/55430#-MJZaTT-9leGh1qmdqe-:-MJhIF6W9w9BRZAHQUWV:btmatmp) ([raw file](https://github.com/cockroachdb/cockroach/blob/f1619dd4f27e6dd14d7e2903a1ffd866c238e86d/pkg/ccl/sqlproxyccl/metrics.go#L29)):*
<details><summary><i>Previously, imjching (Jay Lim) wrote…</i></summary><blockquote>

> ```
> metaCurConnCount
> ```

nit: How about calling this `metaActiveConnCount`?
</blockquote></details>

I've updated the Help of this struct, connects aren't currently being classified as active or inactive. This is just a count.
___
*[pkg/ccl/sqlproxyccl/metrics.go, line 35 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/55430#-MJZaF_BEATnYYKV27Sa:-MJhISvU0Z7F2K740Wd6:b2zd50y) ([raw file](https://github.com/cockroachdb/cockroach/blob/f1619dd4f27e6dd14d7e2903a1ffd866c238e86d/pkg/ccl/sqlproxyccl/metrics.go#L35)):*
<details><summary><i>Previously, imjching (Jay Lim) wrote…</i></summary><blockquote>

> ```
> metaRoutingErrCount
> ```

nit: `s/metaRoutingErrCount/metaRoutingFailedCount`
</blockquote></details>

I pulled the `Err` naming convention from kv's metrics. I'd prefer to stay consistent with the rest of the code base.
___
*[pkg/ccl/sqlproxyccl/metrics.go, line 54 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/55430#-MJZa-qA400eDgRfM_6b:-MJhImGu3qD6m6rLc8QU:bm1brn) ([raw file](https://github.com/cockroachdb/cockroach/blob/f1619dd4f27e6dd14d7e2903a1ffd866c238e86d/pkg/ccl/sqlproxyccl/metrics.go#L54)):*
<details><summary><i>Previously, imjching (Jay Lim) wrote…</i></summary><blockquote>

> ```
> Name:        "proxy.sql.backenddisconnect",
> ```
Why is this metric name the same as `metaBackendDisconnectCount`?

nit: could we also use underscores to split the words, e.g. routing_failed, backend_down, etc?
</blockquote></details>

This was a typo that has been corrected.

It looks like prometheus replaces `.`'s with `_` and metrics from CRDB don't include `_`s.
We could change this to `proxy.sql.disconnect.backend`? I don't feel strongly in either direction.
___
*[pkg/ccl/sqlproxyccl/server.go, line 48 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/55430#-MJZb69vCHIDAo5q7jGa:-MJhK1ad35Nvquyz6MZp:b-bwlq7l) ([raw file](https://github.com/cockroachdb/cockroach/blob/f1619dd4f27e6dd14d7e2903a1ffd866c238e86d/pkg/ccl/sqlproxyccl/server.go#L48)):*
<details><summary><i>Previously, imjching (Jay Lim) wrote…</i></summary><blockquote>

> ```
> registry.AddMetricStruct(proxyMetrics)
> ```

Not needed in this PR, but I'd imagine we want to expose `registry` one day, or at least some form of `AddMetricStruct`. This allows custom implementations, e.g. mtproxy in the context of cockroachdb/cockroach, and sqlproxy in the context of CC to specify its own metrics assuming it does more stuff other than just a regular proxying. For example, sqlproxy could talk to Kubernetes to grab the right routing rule, and maybe that would be a good place to collect metrics as well.
</blockquote></details>

ack and agreed but that's very out of scope for this PR :+1:

**Review by imjching — APPROVED** [review]
:lgtm: once test passes.

> Reviewed 3 of 3 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/55430#-:-MJlnvVTCP4csCpcmHqw:b6eyisx)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @andy-kimball, @imjching, and @knz)

___
*[pkg/ccl/sqlproxyccl/metrics.go, line 29 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/55430#-MJZaTT-9leGh1qmdqe-:-MJhLnXN3KJrILM8hDC8:brxp9sq) ([raw file](https://github.com/cockroachdb/cockroach/blob/f1619dd4f27e6dd14d7e2903a1ffd866c238e86d/pkg/ccl/sqlproxyccl/metrics.go#L29)):*
<details><summary><i>Previously, chrisseto (Chris Seto) wrote…</i></summary><blockquote>

I've updated the Help of this struct, connects aren't currently being classified as active or inactive. This is just a count.
</blockquote></details>

Sounds good.
___
*[pkg/ccl/sqlproxyccl/metrics.go, line 35 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/55430#-MJZaF_BEATnYYKV27Sa:-MJlnKupDowXfUTOniEj:bb6sh2i) ([raw file](https://github.com/cockroachdb/cockroach/blob/f1619dd4f27e6dd14d7e2903a1ffd866c238e86d/pkg/ccl/sqlproxyccl/metrics.go#L35)):*
<details><summary><i>Previously, chrisseto (Chris Seto) wrote…</i></summary><blockquote>

I pulled the `Err` naming convention from kv's metrics. I'd prefer to stay consistent with the rest of the code base.
</blockquote></details>

Ack.
___
*[pkg/ccl/sqlproxyccl/metrics.go, line 54 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/55430#-MJZa-qA400eDgRfM_6b:-MJlnNSCDxu0ASb_TXOD:b-4t8noe) ([raw file](https://github.com/cockroachdb/cockroach/blob/f1619dd4f27e6dd14d7e2903a1ffd866c238e86d/pkg/ccl/sqlproxyccl/metrics.go#L54)):*
<details><summary><i>Previously, chrisseto (Chris Seto) wrote…</i></summary><blockquote>

This was a typo that has been corrected.

It looks like prometheus replaces `.`'s with `_` and metrics from CRDB don't include `_`s.
We could change this to `proxy.sql.disconnect.backend`? I don't feel strongly in either direction.
</blockquote></details>

Discussed offline. We end up with `proxy.err.{routing, backend_down, backend_disconnect, client_disconnect}`.

**Pull request merged**
